Preprocess.

s-alexey opened the pull request:
Here is my suggestion about preprocessing widget. 

Pipeline have been divided into the next stages:
- transformation (lowercase, HTML-parsing)
- tokenization 
- normalization 
- filtering

You can add new issues to [trello board](https://trello.com/b/XtuX6RM8/orange-text)  ([invitation](https://trello.com/invite/b/XtuX6RM8/8d3875140cec3a8792cbbc1e6097e6e0/orange-text))

kernc commented:
So you have started working on your GSoC project, eh? :+1: 
Looks good. A bit ... overwhelming. For me at least. I'm sure @nikicc will love it! :laughing:

kernc commented on orangecontrib/text/string_transformation.py:
```diff
@@ -0,0 +1,72 @@
+    def __call__(self, data):
+        """ Transforms strings in given object.
+        :param data: string or string collection.
+        :type data: str or Iterable
```
Regardless of what it may appear from existing code, we prefer [Napoleon-style docstrings](http://sphinxcontrib-napoleon.readthedocs.org/en/latest/index.html) in new code.

nikicc commented:
Looks great! We were just before merging #21, so I have to check how this conflicts with that.

s-alexey commented:
I have almost completed this PR. Hope you'll like it. 
I rewrite Preprocess and BagOfWords widgets. I also add some useful (from my point of view) things like wrappers and options that allow easily connect third-party classes to widgets. 
Can somebody explain what wrong with Travis and review the code?

nikicc commented:
About Travis, it probably fails due to missing dependencies. I saw that you use `BeautifulSoup` which is missing on Travis. Please, add all required libraries that you added into `setup.py > INSTALL_REQUIRES`.

About the PR in general, we quite like the code! However, it is conflicting with David's work in #21, so we have to decide on how to proceed.

s-alexey commented:
I have rebased and squashed my branch. You can review the current state.

nikicc commented on orangecontrib/text/tests/test_preprocess.py:
```diff
@@ -1,381 +1,226 @@
-
-        # List.
+    sentence = "Human machine interface for lab abc computer applications"
+    corpus = [
```
Could we just load the `deerwester` corpus, like `Corpus.from_file('deerwester')` and use that?

s-alexey replied:
Will change it 👍

nikicc commented on orangecontrib/text/preprocess/tokenize.py:
```diff
@@ -0,0 +1,74 @@
+
+class WordPunctTokenizer(BaseTokenizer):
+    tokenizer = tokenize.WordPunctTokenizer()
+    name = 'Word & punctuation'
```
The `&` sign is not shown in GUI which makes the name a bit confusing.
![screen shot 2016-06-13 at 16 56 00](https://cloud.githubusercontent.com/assets/713026/16011972/c0be8e38-3187-11e6-8771-3bbb25328419.png)

nikicc commented on orangecontrib/text/preprocess.py:
```diff
@@ -1,281 +0,0 @@
-from orangecontrib.text.corpus import Corpus
-
-# Ensure NLTK data is present
-NLTK_DATA = ["wordnet",  "stopwords",  "punkt"]
```
Did you put this somewhere else? I agree this is a bit hackish but we need this somewhere until we figure out a better solution. Without it required NLTK data is not downloaded and preprocessing does not work. Consequently most text widgets also break.

I cleared caches on Travis and rerun the tests to show where is the problem.

s-alexey replied:
Thanks, I will add this lines to `preprocess.__init__.py`

nikicc commented on orangecontrib/text/preprocess/normalize.py:
```diff
@@ -0,0 +1,81 @@
+
+
+class WordNetLemmatizer(BaseNormalizer):
+    name = 'WordNet Limmatizer'
```
A typo. Please change `WordNet Limmatizer` to `WordNet Lemmatizer`.

nikicc commented on orangecontrib/text/preprocess/normalize.py:
```diff
@@ -0,0 +1,81 @@
+
+
+class PorterStemmer(BaseNormalizer):
+    name = 'Porter stemmer'
```
If WordNet Lemmatizer is in camel case then stemmer shoud be also. Same goes for Snowball stemmer.

nikicc commented on orangecontrib/text/preprocess/preprocess.py:
```diff
@@ -0,0 +1,63 @@
+
+
+class Preprocessor:
+    """Holds document processing objects
```
I'm nitpicking here 🙈, but could you add a dot at the end of the sentence.

nikicc commented on orangecontrib/text/tests/test_preprocess.py:
```diff
@@ -1,381 +1,226 @@
+        self.assertRaises(TypeError, tokenizer, 1)
+
+    def test_tokenizer_instance(self):
+        class WhitespaceTokenizer(preprocess.BaseTokenizer):
```
Could we just use `WhitespaceTokenizer` from `tokenize.py` instead?

nikicc commented on orangecontrib/text/tests/test_preprocess.py:
```diff
@@ -1,381 +1,226 @@
+    def test_options(self):
+        tokenizer = preprocess.RegexpTokenizer(pattern='\w')
+        self.assertTrue(hasattr(tokenizer, 'pattern'))
+        tokenizer.pattern = r'\w+'
```
What's the intuition for setting this first to `\w` and the reseting to `\w+`. To check if setters work?

nikicc commented on orangecontrib/text/widgets/owpreprocess.py:
```diff
@@ -1,20 +1,18 @@
-                         QComboBox, QPushButton, QStyle, QApplication,
-                         QFileDialog, QLineEdit, QCheckBox)
-from nltk.corpus import stopwords
+                         QApplication, QLineEdit, QCheckBox)
```
`QLineEdit` seems to be redundant.

nikicc commented on orangecontrib/text/widgets/utils/widgets.py:
```diff
@@ -0,0 +1,111 @@
+        setattr(self.master, self.attr, self.items[i][1])
+
+
+class FileWidget(QWidget):
```
It's very nice that you put this into one common place 👍  Could you also refactor `owloadcorpus.py` to use this instead? I think most of its functionality is the same as here.

nikicc commented:
@s-alexey here are my comments and suggestions. Sorry for not using trello but it was easier for me to just put it here. Also it's probably easier for other developers to follow what's happening. 

- [ ] Filtering, probably also Transformation (and maybe even Tokenization) modules are missing the green ON/OFF button.

- [ ] Green ON/OFF buttons do not work for @ajdapretnar on Windows. Maybe she can describe the issue a bit more into detail but in general the icons are not shown. Though I think the clicking works.

- [ ] Report button should be at its standard location (the lower left corner where we currently have debug info). You can check the Data Table or Predictions widgets for example. It will automatically be placed there if the widget contains the send_report method so you do not need to take care of the Report button. You just need to provide the method.

- [ ] Commit button along with the checkbox should probably also be in the controlArea on left.

- [ ] I know you probably do not agree but I would still prefer if we have `min_df` and `max_df` filtering in this widget (rather than in BoW) since having it here enables the filtering also for all other widgets (topic modelling, word could, ..)

- [ ] The input for regex tokenizer has a confirm button which is at a strange position. To reproduce select Regexp tokneizer and change something in the Pattern field.
![screen shot 2016-06-13 at 17 58 03](https://cloud.githubusercontent.com/assets/713026/16014363/268887f6-3191-11e6-8ec3-cc8bfa7d801a.png)

- [ ] Invalid regexes cause errors (e.g. `sre_constants.error: unexpected end of regular expression`). Regexes should be validated (I see that there is already a method for this in the RegexpTokenizer) and if invalid the warning should be raised.

- [ ] `restore_data` and `export_data` methods are unnecessary complications adopted from Orange's Preprocess widget that really needs them. Here we probably do not need them since the number and order of preprocessors is fixed. Could you try to use normal Settings instead?

s-alexey commented:
@nikicc thanks for reviewing my work 

> Report button should be at its standard location (the lower left corner where we currently have debug info).

Should report button be on top of commit button?

> I know you probably do not agree but I would still prefer if we have min_df and max_df filtering in this widget. 

I agree with that 👍  
I will add `min_df` and `max_df`soon. I'm going to make it something like that: 
```
corpus.dictionary = gensim.corpora.dictionary.Dictionary(corpus.tokens)
corpus.dictionary.filter_extremes(no_below=5, no_above=0.5, keep_n=100000)
```
In this case we could add `keep_n` spin.

> restore_data and export_data methods are unnecessary complications adopted from Orange's Preprocess widget that really needs them. 

I don't use them. Sorry for not removing them properly. 
There is a problem with disable/enable settings, but widget stores in settings preprocess configuration, doesn't it?

ajdapretnar commented:
@s-alexey As @nikicc mentioned, the green on/off buttons somehow don't work on Windows. We've checked and the icons are in the repo where they should be, but they seem not to show on display. However, the functionality of the button is there... Meaning if I click on the area where the button should be, the on/off works.
Here's how it looks like on Windows:

![preprocess-text](https://cloud.githubusercontent.com/assets/12524972/16033591/20c2d8d6-320d-11e6-8ef7-e4f5b421f304.PNG)

One final thing. I'm not sure how the defaults were meant, but I would prefer to set them to some standard preprocessing (lowercase, word&punctuation, porter stemmer, and stopwords in English). The idea is that the user can quickly connect the widgets and get the result. If they want, they can always turn off a part of preprocessing. Could @nikicc please confirm the desired defaults?

nikicc commented:
> Should report button be on top of commit button?

We don't seem to be very consistent about this. Though more widgets have it the way you did it with report button on top of commit one.

When checking this I also find a minor alignment issue. Could you stretch the Commit button a bit to right so it is aligned with the right end of Report button?
<img width="271" alt="screen shot 2016-06-14 at 13 29 11" src="https://cloud.githubusercontent.com/assets/713026/16041111/06fe6c72-3234-11e6-970b-5617faabf38a.png">

> I will add min_df and max_dfsoon. I'm going to do it something like that:

>```
corpus.dictionary = gensim.corpora.dictionary.Dictionary(corpus.tokens)
corpus.dictionary.filter_extremes(no_below=5, no_above=0.5, keep_n=100000)
```

>In this case we could add keep_n spin.

I agree with `keep_n` spin. Though `no_below` and `no_above` should also be inputs in the widget.

> There is a problem with disable/enable settings, but widget stores in settings preprocess configuration, doesn't it?

It would be nice if enable/disable settings would be stored as well. We tried to achieve this some time ago (some messy code resides in [one of my branches](https://github.com/nikicc/orange3-text/commit/65c3d4bbc9f7ae96c0e9c714e5f30cc60438494e)). If I remember correctly, modules should also extend `OWComponent ` and the main widget should use `SettingProvider`. Try to do it this way but if you will have problems let us know and I can check with someone more familiar with Settings.

nikicc commented:
> One final thing. I'm not sure how the defaults were meant, but I would prefer to set them to some standard preprocessing (lowercase, word&punctuation, porter stemmer, and stopwords in English).

These sound like sane defaults. Though the Porter Stemmer might produce some unintuitive tokens but we can disable it later if this will prove annoying.

ajdapretnar commented:
@s-alexey Recent updates look good from a quick inspection. However, I found a strange thing. If I'm in the regex field and I'm typing regex and press 'Enter', a window for loading stopwords list opens. 'Enter' should probably not do anything in this case, as regexes are calculcated upon each change.

@nikicc As for me, Debug sounds fairly strange for an info box. Could we come up with something better? Perhaps 'Preprocessors' or 'Output'? Also the order should be as in the widget itself (normalization before stopwords removal). But these are just cosmetic fixes.

nikicc commented:
> @nikicc As for me, Debug sounds fairly strange for an info box. Could we come up with something better? Perhaps 'Preprocessors' or 'Output'? Also the order should be as in the widget itself (normalization before stopwords removal). But these are just cosmetic fixes.

@ajdapretnar I don't think this is intended to stay. I presume Alexey added this just for debug purposes and this will be removed before merging. After all what would be the purpose of keeping this if it just restates what is selected with combo and checkboxes?

s-alexey commented:
> @s-alexey As @nikicc mentioned, the green on/off buttons somehow don't work on Windows. 

Did the on/off buttons work fine before?

> When checking this I also find a minor alignment issue. Could you stretch the Commit button a bit to right so it is aligned with the right end of Report button?

I couldn't find any solution so far. And it seems other widgets (e.g. Rank, Image Viewer) have the same problem.

ajdapretnar commented:
@s-alexey No, they never worked for me. So strange...

Also, I don't have quite the same issue with the alignment as you two do. See below (it's from Image Viewer, but it's the same elsewhere):
![alignment](https://cloud.githubusercontent.com/assets/12524972/16107350/23fe44aa-339c-11e6-8c3f-5fecee4d418f.PNG)

Another one of the Windows' window manager tricks?

s-alexey commented:
> Another one of the Windows' window manager tricks?

No, I have fixed it yesterday.

ajdapretnar commented:
Yes, but it was never an issue for me on Windows (see the screenshot above*). :smile:

Anyhow, I ticked some to-do boxes @nikicc made. You can review them and tick the ones I missed. I see that ideally we should merge Preprocessing and Twitter by Monday to stay on track. Is this realistic? It seems like you're quite close to finishing, so it would be great to merge everything asap, complete the midterm evaluation and then get on with the second leg of the race. :smile: 

And by the way, I think you're doing a fantastic job. :star2: 

*I guess, I'm getting a bit lost.

nikicc commented:
> Also, I don't have quite the same issue with the alignment as you two do. 

@ajdapretnar but you have a problem with a checkbox 😉

nikicc commented:
@s-alexey `Document frequency range` should also accept values larger that `1.0` and treat them as absolute number of documents. The same way as gensim support absolute numbers as well as fractions.

nikicc commented:
There seems to be a problem with layout. If I disable all preprocessing steps I get this:
![screen shot 2016-06-16 at 12 57 53](https://cloud.githubusercontent.com/assets/713026/16114542/1965ac9c-33c2-11e6-81bc-aa35a16764a8.png)

IMO the auto commit and report buttons should always stick to the lower left corner not matter how many of widgets on the right is enabled/disabled. Also the bounding square for widgets on the right should probably stay the same — i.e. filling the whole widget and not jumping around when modules are enabled/disabled.

You can check the current version of preprocessing. I think we did not have this kind of problems there. Also, FYI I am adding all of those issues to the above task list so we can track more easily.

nikicc commented:
@s-alexey could you rebase to new master? Merging Twitter seems to have caused some issues.

s-alexey commented:
@nikicc the branch is rebased now.

nikicc commented:
@s-alexey I think all issues discussed above are resolved. During this second review I found some more 🙈 and also got some further ideas. Here they are: 

- [ ] Progress bar has some issues. If I put bookexcerpts on input I only see progress 0.25%, 0.5% and 0.75%. Did you forget to multiply by 100?

- [ ] Now that everything is so nicely structured do you think it is possible to improve the granularity of progress reporting? Current division into 25% blocks seems quite coarse. Could we preprocess each document in a loop and report after each document is done or would this introduce to much overhead?

- [ ] Could we also add some additional information into the info box? For example it would be nice to see total and unique number of tokens. For example something like this:

```
Document count: 140
Total tokens: 32345
Unique tokens: 5321
```

- [ ] I suggest we split `Keep most frequent tokens` into a separate independent option in the row below `Document frequency`. Currently it is a bit strange since checkboxes are nested — one must enable document frequency filtering at the beginning of the line first and then also enable second checkbox — and that is not quite obvious at least for me. Also this two options could be used separately so it makes sense to separate the IMO.

- [ ] I know you are using gensim for this but I do not quite understand why this happens. I loaded bookexcerpts, pass them through preprocessor and then to wordcloud. If I enable filtering to 100 most common features some of the most frequent features are filtered out (e.g. `Poly` in example below). So before filtering I see this (check the token frequencies in word cloud):
<img width="1231" alt="screen shot 2016-06-20 at 16 23 01" src="https://cloud.githubusercontent.com/assets/713026/16197768/9c68b17c-3703-11e6-9611-40e9ace1602d.png">
After I enable filtering I get this:
<img width="1232" alt="screen shot 2016-06-20 at 16 23 17" src="https://cloud.githubusercontent.com/assets/713026/16197800/b8959dc4-3703-11e6-8157-8593fbb63b0d.png">
Am I missing something? Where did `Polly` and `Sara` go? If we are filtering by token frequencies shouldn't the most frequent tokens be kept?

- [ ] Could we a tooltip to Document frequency explaining that it accept whole numbers and fractions. 

- [ ] Could you modify defaults to: enable lowercasing, set RegEx (`\w+`) as default tokenizer and enable english stopwords. 

- [ ] In report when we use Lexicon it would be nice if the filename is also printed. Same for Custom stopwords which are now absent from report.

- [ ] Once I select custom stopwords file I cannot remove it unless I disable stopwords alltogether. We should add a way to deselect the file either with None in the dropdown or some other way of deleting this.

- [ ] I do not see a reason to limit the minimal number of Keep most frequent tokens to 10. What if user wants only 5? I agree that in general we want more but keeping the limit at 10 brings not benefits that I could see.

s-alexey commented:
> Progress bar has some issues. If I put bookexcerpts on input I only see progress 0.25%, 0.5% and 0.75%. Did you forget to multiply by 100?

I will fix that. 

> Now that everything is so nicely structured do you think it is possible to improve the granularity of progress reporting? Current division into 25% blocks seems quite coarse. Could we preprocess each document in a loop and report after each document is done or would this introduce to much overhead?

About reporting. Frequency filter requires the whole corpus to learn its length and token frequencies. One of the possibilities is to run all previous stages for each document (possibly, in parallel) and only then run filtering.

> Could we also add some additional information into the info box? For example it would be nice to see total and unique number of tokens. 

I agree with that. Since the same box can be used elsewhere (e.g. in `OWBagOfWords`) I suggest creating an `InfoBox` widget for that purpose. 

> I suggest we split Keep most frequent tokens into a separate independent option in the row below Document frequency.

It will require some redesign since `keep_n` and `min_df` are used in the same function call. I have to think about it.

> I know you are using gensim for this but I do not quite understand why this happens. I loaded bookexcerpts, pass them through preprocessor and then to wordcloud. 

gensim uses document frequencies. Since `Polly` occurs only in 10 documents it filtered out. We should provide a more clear label for this options.

I agree with other suggestions and will add appropriate changes.

nikicc commented:
> About reporting. Frequency filter requires the whole corpus to learn its length and token frequencies. One of the possibilities is to run all previous stages for each document (possibly, in parallel) and only then run filtering.

This sounds good. If filtering would be disabled this would run smoothly. If not we will still have two stages of progress. Feel free to decide on the proportions of progress for each stage as you see fit. About parallelism, I wouldn't bother with that for now.

> I agree with that. Since the same box can be used elsewhere (e.g. in OWBagOfWords) I suggest creating an InfoBox widget for that purpose.

We could just add the text to the existing `Info` widget.

> It will require some redesign since keep_n and min_df are used in the same function call. I have to think about it.

The function call can probably stay the same. We can try to make this work just by modifying `df_changed` and `keep_n_changed` methods if possible.

> We should provide a more clear label for this options.

I agree 👍 

I also found one more peculiarity in min_df filter so I appended it as the last item on to the above TO-DO. Please, check.

Also, let me know when I should check again.

s-alexey commented:
@nikicc can you check the last issue you mentioned?

s-alexey commented:
@nikicc you can check it now.

nikicc commented:
@s-alexey I discovered these problems: 

- [ ] I noticed a huge slowdown comparing to current version of preprocessing. Take `friends-transcripts` data set and use Word & Punctuation, Porter Stemmer and English stopwords. This configuration on current version takes about 40s but on this PR it requires about 5 minutes and a half. That's about 8 times slower. Can you investigate a bit? If the slowdown is the result of precise progress bar than we should find some compromise.

- [ ] Total tokens in info box does not change if I use filtering by Document frequency or Keep most frequent tokens. It should correspond the final filtered version IMO. 

- [ ] I would also add the `Unique tokens: XXX` information, which is actually the vocabulary size. It is useful to know how many features will BoW create and I think these two statistics (number of tokens and vocabulary size) are usually used to reported the size of a corpus.

- [ ] Using stowords in filtering seems to break `Document frequency` and `Keep most frequent tokens` filtering. For example take `bookexcerpts` pass it through preprocessing and then into Word Cloud (which comes in handy since it prints the number of unique words). If I do not use English stopwords and set to Keep to 50 this works OK and WordCloud also shows 50 words. If I also enable English Stopwords and keep Keep=50 I see 11592 words in WordCloud but it should show 50. Seems like by enabling stopwords  and running preprocessing once Document and Keep filtering are broken forever, i.e. they do not work any more even if I disable stopwords.

- [ ] Rename `Keep most frequent tokens:` to something like `Keep top tokens by document frequency:`

- [ ] Could we a tooltip to Document frequency explaining that it accept whole numbers and fractions.

s-alexey commented:
> I noticed a huge slowdown comparing to current version of preprocessing. 

Thank for noticing, it's really important. Yes, it seems to be caused by the reporting. If you disable reporting its runs much faster (for me it runs even faster than the current implementation). As a possible solution, we could change report frequency, for example, report every 100 documents. This will speed up preprocessing significantly. I have made changes, you can test it now.

>  Using stowords in filtering seems to break Document frequency and Keep most frequent tokens filtering.

Fixed.

I have added `Corpus.dictionary` attribute that allows us to get unique tokens count.

nikicc commented:
> I have made changes, you can test it now.

This is much better 👍

nikicc commented:
>> Using stowords in filtering seems to break Document frequency and Keep most frequent tokens filtering.

>Fixed.

Still doesn't always work. If I first run filtering with Keep=100 and without stopwords it works:
![screen shot 2016-06-23 at 12 33 36](https://cloud.githubusercontent.com/assets/713026/16300172/d34d8478-393e-11e6-9e9b-4fce14eb3921.png)

But if I enable stopwords now it still breaks:
![screen shot 2016-06-23 at 12 33 45](https://cloud.githubusercontent.com/assets/713026/16300205/fc29c94c-393e-11e6-9169-c3c20042d39d.png)

nikicc commented:
@s-alexey we created a [minimal working example](https://gist.github.com/nikicc/468543a494ea1a14a4a82e605a9d99de) with a button that works on mac and windows. Can you replace QCheckBox with QPushButton as shown in example?

Just for convenience these are currently open issues:

- [ ] Stopwords filtering breaks `Document frequency` and `Keep top tokens`.
- [ ] Replace ON/OFF icon.
- [ ] Fix tests.

s-alexey commented:
> Still doesn't always work. If I first run filtering with Keep=100 and without stopwords it works

@nikicc sorry, my fault. I have fixed method changing process, so it should work now.

s-alexey commented:
@ajdapretnar Can you check whether On/Off button is visible on Windows now?

ajdapretnar commented:
@s-alexey I've tested from my home laptop and I had to solve of bunch of dependency issues, so I cannot guarantee the result. But now I can see the icons, only that they're green when the box is disabled and grey when it's enabled. Should that be the case?

<img width="544" alt="preprocess-text" src="https://cloud.githubusercontent.com/assets/12524972/16316458/ff3a0fa0-3986-11e6-84c2-47b4e4aaf4d0.PNG">

I'll test tomorrow morning from work as well.

nikicc commented:
> Should that be the case?

No. They should be exactly the opposite IMO.

s-alexey commented:
Sorry for that :)

nikicc commented:
@s-alexey would you also consider adding something like `self.setStyleSheet('border:none')` to `OnOffButton`? Currently, I get this not so pretty square box around the button when clicked.

<img width="81" alt="screen shot 2016-06-23 at 23 08 24" src="https://cloud.githubusercontent.com/assets/713026/16320345/8610c126-3997-11e6-8235-d49231eb688d.png">

nikicc commented:
Did you disable reporting from `process_document` on purpose? Currently there is no reporting between 0% and 80%. I suggest we turn `report=True` in `process_document `.

s-alexey commented:
> @s-alexey would you also consider adding something like self.setStyleSheet('border:none') to OnOffButton? 
👍 It's look much pretier now

> Did you disable reporting from process_document on purpose? Currently there is no reporting between 0% and 80%. I suggest we turn report=True in process_document.

Will fix that.

nikicc commented:
Looks good to me. I will wait for @ajdapretnar to confirm that it works on windows, check everything once again in the morning and then merge. @s-alexey very nice work 👍

ajdapretnar commented:
@s-alexey @nikicc First thing in the morning I checked this and it's great. Amazing job! Let's merge! :horse_racing:
